worker/backend/container.go

```diff
@@ -1,39 +1,68 @@
 package backend
 
 import (
+	"context"
 	"io"
+	"syscall"
 	"time"
 
 	"code.cloudfoundry.org/garden"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/errdefs"
 )
 
-type Container struct{
-	handle string
+type Container struct {
+	ctx                 context.Context
+	containerdContainer containerd.Container
 }
 
-// func NewContainer() Container {
-// 	return Container{}
-// }
+func NewContainer(ctx context.Context, container containerd.Container) garden.Container {
+	return &Container{ctx, container}
+}
 
 var _ garden.Container = (*Container)(nil)
 
-func (c *Container) Handle() (handle string) { return c.handle }
+// Handle returns a container's ID. This is the handle provided to the backend when it created the container.
+//
+func (c *Container) Handle() string { return c.containerdContainer.ID() }
 
 // Stop stops a container.
 //
-// If kill is false, garden stops a container by sending the processes running inside it the SIGTERM signal.
-// It then waits for the processes to terminate before returning a response.
-// If one or more processes do not terminate within 10 seconds,
-// garden sends these processes the SIGKILL signal, killing them ungracefully.
-//
-// If kill is true, garden stops a container by sending the processing running inside it a SIGKILL signal.
-//
-// It is possible to copy files in to and out of a stopped container.
-// It is only when a container is destroyed that its filesystem is cleaned up.
+// If kill is false, the backend stops a container by sending the processes running inside it the SIGTERM signal.
+// If kill is true, the processing is ungracefully terminated with a SIGKILL signal.
 //
 // Errors:
 // * None.
-func (c *Container) Stop(kill bool) (err error) { return }
+func (c *Container) Stop(kill bool) error {
+	task, err := c.containerdContainer.Task(c.ctx, nil)
```

Comment:
If I understood correctly, the idea here is that `ctx` will come with a timeout, which would enforce a total duration to wait for the process to be stopped - is that right?

If so, I don't think that's what actually going on under the hood - when you create a context with a particular timeout, "the clock starts ticking": a deadline is set at that very exact moment, thus, time is already "running out" from that very moment.

For instance, consider the following snippet:

```go
func sleepHalfTheTime(ctx context.Context, total time.Duration) {
	time.Sleep(total >> 1)
}

func doSomethingUntilContextExpieration(ctx context.Context) {
	start := time.Now()
	defer func() {
		fmt.Printf("ELAPSED: %s\n", time.Now().Sub(start))
	}()

	<-ctx.Done()
}

func main() {
	total := 10 * time.Second

	ctx, _ := context.WithTimeout(context.TODO(), 10*time.Second)

	sleepHalfTheTime(ctx, total)
	doSomethingUntilContextExpieration(ctx)
}
```

this would result in:


```
ELAPSED: 5.000082213s
```

see [`WithTimeout` using `WithDeadline`](https://github.com/golang/go/blob/0ad368675bae1e3228c9146e092cd00cfb29ac27/src/context/context.go#L496)

```go
func WithTimeout(parent Context, timeout time.Duration) (Context, CancelFunc) {
	return WithDeadline(parent, time.Now().Add(timeout))
}
```

Reply:
I think that might be a topic how a garden container's `Stop()`  is called.
There is a bool parameter to indicate how the container is stoped. Can we assume the kill process is always `SIGTERM`, then `SIGKILL` here?
I guess the assumption may be:
- the caller is responsible for the context setup, like `WithTimeout` or `WithCancel`. If the context is a timeout context, the caller is responsible for the timeout duration as well.
- the caller is responsible for calling the `Stop()` with `true` or `false`. and here is example how garden stop() a container: https://github.com/cloudfoundry/garden/blob/044d7ac89d98e7ae70aacb86fce9ff434364bfe8/server/request_handling.go#L196

WDYT?